compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,519 @@
+    pos_bol: 0,
+    pos_cnum: stmsEndtChar,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmtStartLine,
+    pos_bol: 0,
+    pos_cnum: stmsStartChar,
+  };
+
+  let location: Grain_parsing.Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: true,
+  };
+  get_comments_inside_location(~location, comments);
+};
+
+let get_comments_enclosed_and_before_location =
+    (
+      ~loc1: Grain_parsing.Location.t,
+      ~loc2: Grain_parsing.Location.t,
+      comments: list(Grain_parsing.Parsetree.comment),
+    ) => {
+  let (_, locStartLine, locStartChar, _) =
+    Locations.get_raw_pos_info(loc1.loc_start);
+  let (_, stmtStartLine, stmsStartChar, _) =
+    Locations.get_raw_pos_info(loc2.loc_start);
+
+  // invert the request to look inside the location in the gap
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: locStartLine,
+    pos_bol: 0,
+    pos_cnum: locStartChar,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmtStartLine,
+    pos_bol: 0,
+    pos_cnum: stmsStartChar,
+  };
+
+  let location: Grain_parsing.Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: true,
+  };
+
+  get_comments_inside_location(~location, comments);
+};
+
+let get_comments_to_end_of_enclosing_location =
+    (
+      ~wrapper: Grain_parsing.Location.t,
+      ~location: Grain_parsing.Location.t,
+      comments: list(Grain_parsing.Parsetree.comment),
+    ) => {
+  let (_, wrapEndLine, wrapEndChar, _) =
+    Locations.get_raw_pos_info(wrapper.loc_end);
+  let (_, locEndLine, locEndChar, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: locEndLine,
+    pos_bol: 0,
+    pos_cnum: locEndChar,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: wrapEndLine,
+    pos_bol: 0,
+    pos_cnum: wrapEndChar,
+  };
+
+  let location: Grain_parsing.Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: true,
+  };
+  let cmts = get_comments_inside_location(~location, comments);
+
+  cmts;
+};
+
+let get_comments_from_start_of_enclosing_location =
+    (
+      ~wrapper: Grain_parsing.Location.t,
+      ~location: Grain_parsing.Location.t,
+      comments: list(Grain_parsing.Parsetree.comment),
+    ) => {
+  let (_, wrapStartLine, wrapStartChar, _) =
+    Locations.get_raw_pos_info(wrapper.loc_start);
+  let (_, locStartLine, locStartChar, _) =
+    Locations.get_raw_pos_info(location.loc_start);
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: wrapStartLine,
+    pos_bol: 0,
+    pos_cnum: wrapStartChar,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: locStartLine,
+    pos_bol: 0,
+    pos_cnum: locStartChar,
+  };
+
+  let location: Grain_parsing.Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: true,
```

Comment:
Is this really a ghost? It does exist in the file, doesn't it?

Reply:
It does.  I made it true because I manufactured it, but it does exist.  Switched to false everywhere